### ISSUE TITLE
runWarnings format difficult for pragmatic usage

### ISSUE THREAD

**Issue by benschwarz** [GoogleChrome/lighthouse#11352]
I have two concerns with `runWarnings` specifically around pragmatic usage: 

* `runWarnings` contains both errors (eg, page hung / non-successful request on main document resource) and warnings (eg, CPU throttling might be bad)
* The errors and warnings contained within `runWarnings` are strings. There's no handle to be able to identify one warning/error from another, without searching the string. 

**Feature request summary**

`runWarnings` to be changed to an object. It should contain `code` and `message`. Just like `runtimeError`. 

**What is the motivation or use case for changing this?**

Node.js/API/PageSpeed users of Lighthouse will be able to handle errors/warnings reliably.

**Comment by paulirish** [comment]
We're thinking we can add a `runtimeWarnings` field on the LHR which would contain error/code combos like you described. We'd keep `runWarnings` around until we remove it in a major release/breaking change.

sgty?

**Comment by benschwarz** [comment]
@paulirish that sounds good. On a bit of review we also noted that audits also do not have structured error messages: 

```
"audits": {
    "metrics": {
      "id": "metrics",
      "title": "Metrics",
      "description": "Collects all available metrics.",
      "score": null,
      "scoreDisplayMode": "error",
      "rawValue": null,
      "errorMessage": "Something went wrong with recording the trace over your page load. Please run Lighthouse again. (NO_TRACING_STARTED)"
    }
  }
```

What do you think about expanding this issue to add `runtimeWarnings` to `audits`?

**Comment by patrickhulce** [comment]
> What do you think about expanding this issue to add runtimeWarnings to audits?

It's a much larger scope to formalize every error that can be thrown in every audit. There is a small class of page-level warnings that are similar in spirit to `runtimeError` which makes sense to formalize. 

Is it mostly the errors that are already known you'd like as an object in `errorMessage`?

**Comment by benschwarz** [comment]
> Is it mostly the errors that are already known you'd like as an object in errorMessage?

Yep. At the moment they need to be parsed to be machine readable. 
Ideally every existing error message with a `code` would have a structured object: 

```
{
  "code": "NO_TRACING_STARTED",
  "message": "Something went wrong with recording the trace over your page load. Please run Lighthouse again.",
}
```

**Comment by spideyashith** [comment]
Hi! I'd like to work on this issue.

From the discussion above, my understanding of the scope is:
- Add a new `runtimeWarnings` field to the LHR that contains structured `{ code, message }` objects, following the same pattern as `runtimeError`.
- Keep the existing `runWarnings` field as-is for backward compatibility, per @paulirish's comment.
- This is scoped to page-level warnings (not expanding into per-audit `errorMessage` structuring, which @patrickhulce noted is a separate, larger scope).

I'll dig into the codebase to find where `runWarnings` and `runtimeError` are currently defined/populated and follow up here once I have a plan, before opening a PR. Let me know if I'm missing anything or if this scope has changed since 2021.

**Comment by spideyashith** [comment]
Quick update: I've added the `runtimeWarnings?: Array<{ code: string; message: string }>` type to `types/lhr/lhr.d.ts` (type-checks cleanly) and mapped out where page-level warnings currently get pushed into `LighthouseRunWarnings`.

There are 5 call sites today. Three are self-contained (one warning type each) and straightforward to pair with a code:
- `navigation-runner.js` (pageLoadError.friendlyMessage) — can reuse the existing `pageLoadError.code` (e.g. NO_FCP, PAGE_HUNG, TARGET_CRASHED) directly
- `timespan-runner.js` (navigation-detected warning) — would assign a new code
- `link-elements.js` (header parse warning) — would assign a new code

Two others (`getEnvironmentWarnings()` and `prepareTargetForNavigationMode()`) each bundle several distinct warning types internally (clear-data, clear-cache, storage warnings), so giving those proper individual codes looks like a bigger follow-up task rather than part of this first PR.

Plan: implement `runtimeWarnings` for the three self-contained sources now, leave `runWarnings` untouched for backward compatibility, and note the other two sources as follow-up work in the PR description. Let me know if that split makes sense or if you'd rather I take a different approach.

**Comment by spideyashith** [comment]
I've implemented the scoped changes and opened #17208 for review. 

All unit tests and type-checks are passing cleanly with the existing `runWarnings` preserved for backward compatibility. Whenever you get a chance to take a look, feedback is welcome!

**Comment by connorjclark** [comment]
@benschwarz if this feature still interests you, could you help review the PR please?
